### ISSUE TITLE
Inlining of deps from empty chunks with preserveModules doesn't check the moduleSideEffects flag

### ISSUE THREAD

**Issue by dmichon-msft** [rollup/rollup#3359]
- Rollup Version: 1.29.1
- Operating System (or Browser): Windows 10
- Node Version: 10.18.0

Search terms:
treeshaking
moduleSideEffects
preserveModules
empty imports

Related:
#3057 
#3109 

### How Do We Reproduce?

Using the module layout here:
[REPL](https://rollupjs.org/repl/?version=1.29.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMGElMjAlN0QlMjBmcm9tJTIwJy4lMkZpbmRleCclM0IlNUNuaW1wb3J0JTIwJTdCJTIwZCUyMCU3RCUyMGZyb20lMjAnLiUyRm9uZSclM0IlNUNuJTVDbmNvbnNvbGUubG9nKGElMjAlMkIlMjBkKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmluZGV4LmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMCU3QiUyMGElMjAlN0QlMjBmcm9tJTIwJy4lMkZhJyUzQiU1Q25leHBvcnQlMjAlN0IlMjBiJTIwJTdEJTIwZnJvbSUyMCcuJTJGYiclM0IlNUNuZXhwb3J0JTIwJTdCJTIwYyUyMCU3RCUyMGZyb20lMjAnLiUyRmMnJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmEuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwY29uc3QlMjBhJTIwJTNEJTIwMSUzQiUyMiU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJiLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNvbnN0JTIwYiUyMCUzRCUyMDIlM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyYy5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjb25zdCUyMGMlMjAlM0QlMjAzJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm9uZS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjAlN0IlMjBiJTIwJTdEJTIwZnJvbSUyMCcuJTJGaW5kZXgnJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyMGQlMjAlM0QlMjBiJTIwJTJCJTIwNCUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlc20lMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==)
Specify `preserveModules: true` and `treeShaking.moduleSideEffects: false`

I understand the contract of the `moduleSideEffects` flag to mean, for a given module:
- `true`: All modules imported later than this module may have an implicit dependency on this module, and therefore this module must not be moved to be later in the load order (but may potentially be moved earlier)
- `false`: No modules have an implicit dependency on this module, and thus it may be safely moved later in the load order

### Expected Behavior
```js
// main.js
import { a } from '../a.js';
import { d } from '../other.js';

console.log(a + d);
```

### Actual Behavior
```js
// main.js
import { a } from '../a.js';
import '../b.js';
import { d } from '../other.js';

console.log(a + d);
```

Addressing this may be as simple as introducing a check in `inlineChunkDependencies`:
```diff
                if (!dep.isEmpty) {
+                   // In preserveModules mode, only inline the dep if it has side effects (being used elsewhere is not a side effect)
+                   if (!this.graph.preserveModules || !dep.facadeModule || dep.facadeModule.moduleSideEffects) {
                        this.dependencies.push(dep);
+                   }
                }
```

Edit: added info about `moduleSideEffects` expectations
Edit: added link to 3109

**Comment by lukastaegert** [comment]
The actual solution I implemented was a little more complicated as pruning dependencies should ideally happen on module level, not on chunk level. Nevertheless, there is a pending fix now: #3369 !

**Comment by lukastaegert** [comment]
Solved in rollup@2.0.0
